homeassistant/components/utility_meter/sensor.py

```diff
@@ -153,11 +157,26 @@ def __init__(
         self._tariff = tariff
         self._tariff_entity = tariff_entity
 
+    def start(self, unit):
+        """Initialize unit and state upon source initial update."""
+        self._unit_of_measurement = unit
+        self._state = 0
+        self.async_write_ha_state()
+
     @callback
     def async_reading(self, event):
         """Handle the sensor state changes."""
         old_state = event.data.get("old_state")
         new_state = event.data.get("new_state")
+
+        if self._state is None and new_state.state:
+            # First state update initializes the utility_meter sensors
+            source_state = self.hass.states.get(self._sensor_source_id)
+            for sensor in self.hass.data[DATA_UTILITY][self._parent_meter][
```

Comment:
Please use our dispatch helper to signal the entity instead of storing the entities in a shared container. Accessing entities from outside is fragile.

Reply:
Such a solution was implemented previously in this PR, I think it's @dgomes may have force-pushed so it's no longer visible. Doing it via signaling instead introduces a race which we must then guard against and the whole things becomes over complicated.

I think this is a better solution in this case, the children entities of a utility_meter can't change unless the configuration is changed and the configuration is reloaded.

Reply:
What's the race?

Reply:
The race is that the utility meter's sub sensors may race against each other.

Let's say we have a utility_meter with two tariffs, sensor.tarif_1 and sensor.tarif_2, both based on sensor sensor.energy_meter

1. sensor.energy_meter has a state update, unit_of_measurement = "kWh"
2. sensor.tariff_1 gets an initial state and resets itself with unit_of_measurement = kWh + dispatches a signal for the other entities to also reset
3. sensor.energy_meter has a state update, unit_of_measurement = "rabbits"
4. tariff is changed, sensor.tariff_2 gets its initial state and resets itself with unit_of_measurement = "rabbits + dispatches a signal for the other entities to also reset
5. sensor.tariff_2 receives the reset signal from sensor.tariff_1
6. sensor.tariff_1 receives the reset signal from sensor.tariff_2

The example is maybe not realistic, but the race is there.

Reply:
It looks like the problem lies in the source of the signal instead of the fact that we're using signals. Why does one entity cause other entities to reset? I'd expect there to be a separate coordinator that decides this.

Reply:
Before this PR:
- `utility_meter` set its state to 0, with no `unit_of_measurement` and no `device_class` when initialized
- when receiving a valid state from the source sensor, the `unit_of_measurement` and `device_class` are both updated

This behavior was not wanted, because 0 is a perfectly valid state, but the `unit_of_measurement` and `device_class` were not yet known.

With this PR, we wait for a valid state from the source sensor before initializing the utility_meter's state to 0 and setting `device_class` and `unit_of_measurement`

Reply:
Right. I'm saying I think we can avoid the race if we move the coordination of resetting or updating the sensors for the same source to a separate single coordinator. It's possible I'm missing some detail why that's not possible.

Reply:
The requirement to have the `unit_of_measurement` always set led to these developments, previously the tariff sensor would get its `unit_of_measurement` set whenever it started collecting independently. There was no need for any coordination (and therefore non existed)

This current patch provides IMHO such coordination because each sensor can share it's `unit_of_measurement` with the remaining sensors.

There is added benefit: the utility_meter can now track all it's child sensors, if I need to sum all the tariffs I can programatically get all child tariffs from the utility_meter (that will be a future PR).

Reply:
> we can avoid the race if we move the coordination of resetting or updating the sensors for the same source to a separate single coordinator.

Now I understand what you mean Martin! That would indeed be a nicer solution; the uility_meter parent tracks the source sensor and updates its children.

Reply:
👍